### PR TITLE
Job matix

### PR DESCRIPTION
**Pull request by jovdipp**

### CODE REVIEW

File: src/emqx_ct_helpers_docker.erl [hunk 1]
```diff
@@ -51,45 +53,29 @@ set_env_file(Env) ->
...
+                    ok
+            after 1 ->
+                ok
+            end,
+            ExitCode = receive
+                           { Port, { exit_status, Code } } ->
+                               Code
+                       end,
+            { ExitCode, lists:flatten( Sofar) }
+    end.
```
Comment (zmstone) [inline src/emqx_ct_helpers_docker.erl:81]: Could you configure IDEA not to delete the last LF, rather always ensure a LF at EOF?

File: src/emqx_ct_helpers_docker.erl [hunk 2]
```diff
@@ -51,45 +53,29 @@ set_env_file(Env) ->
...
-parse(Lines0) ->
-    %% last line is alwyas a \n
-    [$\n | Lines] = lists:reverse(Lines0),
-    %% second last line is $?
-    {LastLine, OutputLines} = lists:splitwith(fun (C) -> C =/= $\n end, Lines),
-    {list_to_integer(lists:reverse(LastLine)), lists:reverse(OutputLines)}.
-
--ifdef(TEST).
--include_lib("eunit/include/eunit.hrl").
+sh(Cmd) ->
```
Comment (zmstone) [inline src/emqx_ct_helpers_docker.erl:56]: thank you!

File: src/emqx_ct_helpers_docker.erl [hunk 3]
```diff
@@ -16,7 +16,7 @@
 
 -module(emqx_ct_helpers_docker).
 
--export([docker_ip/2, compose/5, stop/1]).
+-export([docker_ip/2, compose/5, stop/1, remove/1]).
```
Comment (zmstone) [inline src/emqx_ct_helpers_docker.erl:19]: sorry I still don't have the full picture
is the usage always like stop then remove? if yes, we do not need the stop API,
a `-f` option for `docker rm` will force it.
i.e. only need `remove/1` ?

File: src/emqx_ct_helpers_docker.erl (hunk 1 above)
Comment (jovdipp) [reply src/emqx_ct_helpers_docker.erl:81]: yeah

File: src/emqx_ct_helpers_jobs.erl [hunk 4]
```diff
@@ -81,7 +81,15 @@
 	JobsConfig = [ { job, CurrentJob }, { all_jobs, Jobs } ],
 	ConfigUpdated = [ { ?JOBS_MATRIX_CONFIG, JobsConfig } | Config ],
 	case is_job_node() of
-		true -> override_function(FuncExists, Suite, ?CT_INIT_PER_SUITE, [ConfigUpdated], Config);
+		true ->
+			try override_function(FuncExists, Suite, ?CT_INIT_PER_SUITE, [ConfigUpdated], Config) of
```
Comment (zmstone) [inline src/emqx_ct_helpers_jobs.erl:85]: move it to a function ?

File: src/emqx_ct_helpers_docker.erl (hunk 3 above)
Comment (jovdipp) [reply src/emqx_ct_helpers_docker.erl:19]: yes, the current usage of in the only Suite using it is stopping & then removing. The module as titled "...helpers_docker" is meant to provide some reasonable re-usable functions for other who will be using the module in more nuanced ways from other Suites.
Adding an additional force_remove function would be good for my current usage.

File: src/emqx_ct_helpers_jobs.erl (hunk 4 above)
Comment (jovdipp) [reply src/emqx_ct_helpers_jobs.erl:85]: The override_function already has a try catch for better exception reporting, but this specific implementation is due to needing to end_per_job to trigger despite any failure in init_per_suite, as such is unique.

Comment (jovdipp) [reply src/emqx_ct_helpers_jobs.erl:85]: yes, I'll expand on the existing try catch in the override_function

File: src/emqx_ct_helpers_docker.erl (hunk 1 above)
Comment (jovdipp) [reply src/emqx_ct_helpers_docker.erl:81]: done

File: src/emqx_ct_helpers_docker.erl (hunk 3 above)
Comment (jovdipp) [reply src/emqx_ct_helpers_docker.erl]: Added

**Review by zmstone — APPROVED** [review]

**Pull request merged**